controller/cmd/service-mirror/cluster_watcher.go

```diff
@@ -0,0 +1,782 @@
+	// reconcile. Most importantly we need to keep track of exposed ports
+	// and gateway association changes.
+	RemoteServiceUpdated struct {
+		localService   *corev1.Service
+		localEndpoints *corev1.Endpoints
+		remoteUpdate   *corev1.Service
+		gatewayData    *gatewayMetadata
+	}
+
+	// RemoteServiceDeleted when a remote service is going away or it is not
+	// considered mirrored anymore
+	RemoteServiceDeleted struct {
+		Name      string
+		Namespace string
+	}
+
+	// RemoteGatewayDeleted is observed when a service that is a gateway to at least
+	// one already mirrored service is deleted
+	RemoteGatewayDeleted struct {
+		gatewayData       *gatewayMetadata
+		affectedEndpoints []*corev1.Endpoints
+	}
+
+	// RemoteGatewayUpdated happens when a service that is a gateway to at least
+	// one already mirrored service is updated. This might mean an IP change,
+	// incoming port change, etc...
+	RemoteGatewayUpdated struct {
+		newPort              int32
+		newEndpointAddresses []corev1.EndpointAddress
+		gatewayData          *gatewayMetadata
+		newResourceVersion   string
+		affectedServices     []*corev1.Service
+	}
+
+	// ConsiderGatewayUpdateDispatch is issued when we are receiving an update for a
+	// service but we are not sure that this is a gateway. We need to hit the local
+	// API and make sure there are services that have this service as a gateway. Since
+	// this is an operation that might fail (a glitch in the local api connectivity)
+	// we want to represent that as a separate event that can be requeued
+	ConsiderGatewayUpdateDispatch struct {
+		maybeGateway *corev1.Service
+	}
+
+	// ClusterUnregistered is issued when the secret containing the remote cluster
+	// access information is deleted
+	ClusterUnregistered struct{}
+
+	// OprhanedServicesGcTriggered is a self-triggered event which aims to delete any
+	// orphaned services that are no longer on the remote cluster. It is emitted every
+	// time a new remote cluster is registered for monitoring. The need for this arises
+	// because the following might happen.
+	//
+	// 1. A cluster is registered for monitoring
+	// 2. Services A,B,C are created and mirrored
+	// 3. Then this component crashes, leaving the mirrors around
+	// 4. In the meantime services B and C are deleted on the remote cluster
+	// 5. When the controller starts up again it registers to listen for mirrored services
+	// 6. It receives an ADD for A but not a DELETE for B and C
+	//
+	// This event indicates that we need to make a diff with all services on the remote
+	// cluster, ensuring that we do not keep any mirrors that are not relevant anymore
+	OprhanedServicesGcTriggered struct{}
+
+	gatewayMetadata struct {
+		Name      string
+		Namespace string
+	}
+)
+
+func (rcsw *RemoteClusterServiceWatcher) getLocalService(ns, name string) (*corev1.Service, error) {
+	svc, err := rcsw.localAPIClient.Svc().Lister().Services(ns).Get(name)
+	if err != nil {
+		return nil, err
+	}
+	return svc.DeepCopy(), nil
```

Comment:
I am using deep copy because the services I get are out of the cache in the Informer. I do not want to modify them by acceident and change its state.

Reply:
Do we ever modify the services?  I think it would be better practice to be deliberate and specific about where we do copies.  I.e. don't do blanket copies like this but instead always do a copy immediately before modifying the object.

Reply:
Is this being blanket? It seems like we're copying when we take something out of the infromer, generally, which seems like a decent strategy?

To put it the other way: When would you copy? Is it actually a better practice to copy when mutating? What if there's no possibility of that data being shared?

Reply:
I don't actually see any place where we are mutating.  As far as I know, we're only reading these objects and shouldn't need to copy at all.  

To put it in rust terms: I think of this as an immutable borrow and if we ever need to do mutations we can always call clone() at that time.

Reply:
> To put it in rust terms: I think of this as an immutable borrow and if we ever need to do mutations we can always call clone() at that time.

Ecept that we can mutate it and nobody is going to complain :D

Reply:
Right, to my thinking: since we are taking a shared data structure and returning it, we can't assume how it will be used (mutably or not) so copying seems like the safest choice until we have some concrete performance motive to avoid copying.